2_process/src/munge_GFv1_catchments.R

```diff
@@ -0,0 +1,147 @@
+  intsct_hrus$length_percent = 100 * (sf::st_length(intsct_hrus)/intsct_hrus$length)
+  
+  # Return attribute `hru_segment` for the HRU that has the greatest overlap with prms_line
+  hru_segment <- intsct_hrus$hru_segment[which.max(intsct_hrus$length_percent)]
+  
+  return(hru_segment)
+  
+}
+
+
+
+return_nhdv2_cat_shps <- function(prms_line,segs_w_comids){
+  #' 
+  #' @description This function takes a data table of PRMS segment ID's and
+  #' its contributing NHDv2 tributary catchments and returns those catchments
+  #' as an sf object
+  #' 
+  #' @param prms_line sf object containing one PRMS river segment of interest
+  #' prms_line must contain variables subsegid and geometry
+  #' @param segs_w_comids data frame containing the PRMS segment ids and the contributing COMID's
+  #' segs_w_comids must contain variables PRMS_segid and COMID
+  #' 
+  
+  # Subset segs_w_comids to only include the PRMS segment of interest
+  nhd_cats <- segs_w_comids %>%
+    filter(PRMS_segid == prms_line$subsegid)
+  
+  # Use nhdplusTools to retrieve the catchment polygons for COMIDs of interest
+  nhd_cats_sf <- nhdplusTools::get_nhdplus(comid = c(nhd_cats$COMID),realization = "catchment",t_srs = 5070) %>%
+    # dissolve boundaries between individual NHDv2 catchments
+    sf::st_union() %>%
+    # convert back to sf data.frame object and add the segment ID
+    sf::st_as_sf() %>%
+    mutate(PRMS_segid = unique(prms_line$subsegid)) %>%
+    rename(geometry = x)
+  
+  return(nhd_cats_sf)
+  
+}
+
+
+
+munge_GFv1_catchments <- function(prms_lines,prms_hrus,segs_w_comids,crs_out = 4326){
+  #' 
+  #' @description This function munges the PRMS HRU's (~ the PRMS catchments) so that every
+  #' PRMS segment in the Delaware River Basin has at least one corresponding HRU and so that 
+  #' the catchment boundaries are adjusted for 3 NHGFv1 segments that were split in the 
+  #' `delaware-model-prep` pipeline: segid's 3_1, 3_2, 8_1, 8_2, 51_1, 51_2 
+  #' (https://github.com/USGS-R/delaware-model-prep) 
+  #' 
+  #' @param prms_lines sf object containing the PRMS river segments for the DRB
+  #' prms_lines must contain variables subsegid and geometry
+  #' @param segs_w_comids data frame containing the PRMS segment ids and the contributing COMID's
+  #' segs_w_comids must contain variables PRMS_segid and COMID
+  #' @param prms_hrus sf (multi)polygon containing the HRU's from the NHGFv1
+  #' @param crs_out numeric EPSG code indicating desired crs. Defaults to EPSG: 4326, WGS84.
+  #
+  
+  # 1. Fill PRMS lines data frame so that every segment has a value for the attribute
+  # `hru_segment`, which can be used as a key to join each segment to its corresponding HRU polygons.
+  prms_hrus_filled <- prms_lines %>%
+    # Transform prms_lines into a list with a data frame for each PRMS segment (represented by a unique subsegid) 
+    split(.,.$subsegid) %>%
+    # For each PRMS segment, find the HRU with the greatest overlap and add as an attribute
```

Comment:
How long does it take to run this intersection? It looks like it's currently applied to all segments, but you could run it for only the segments with missing HRUs.

Reply:
You're right, the code currently finds the intersecting HRU's for all segments. It takes less than 2 minutes to run on my computer, but I agree that we're doing more computations than we need here. I modified the `lapply` function starting on line 90 to only find the intersecting HRU if the PRMS segment is missing HRUs, otherwise use the attribute `subsegseg`, which matches `hru_segment` within the HRU polygons data table. This step now takes 15 seconds, thanks for the nudge!